Initial basic support for SF32LB52x SoCs

gmarull opened the pull request:
SF32LB52x is a family of SoCs from SiFli Technologies based on the Arm Star-MC1 core (Cortex-M33 compatible).

For more details, see:
https://wiki.sifli.com/en/silicon/index.html
https://wiki.sifli.com/en/hardware/SF32LB520-3-5-7-HW-Application.html
https://wiki.sifli.com/en/hardware/SF32LB52B-E-G-J-HW-Application.html

This PR adds elementary support for the platform; that is, it boots "hello world" printing on UART. Any additions or improvements will come later.

As you'll see, a new HAL is added, which can also be used by users to build stuff at the application level, bypassing Zephyr functionality, or where Zephyr interfaces are missing. Because SiFli HAL is particularly "fat" and does many things under the hood, I've tried to stick to simple register writes where applicable, as it leads to a much better experience, performance, and readability.

This will be used to build the Pebble Time 2 stage 0 bootloader.

Board: https://builds.zephyrproject.io/zephyr/pr/95476/docs/boards/sifli/sf32lb52_devkit_lcd/doc/index.html

Fixes #95312

gmarull committed:
dts: bindings: add sifli vendor prefix

Add `sifli` for SiFli Technologies(Nanjing) Co., Ltd.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: sifli: introduce SF32LB52x

SF32LB52x is a SoC from SiFli Technologies(Nanjing) Co., Ltd, based on
Arm Star-MC1 core (Cortex-M33 compatible).

For more details, see:
https://wiki.sifli.com/en/hardware/SF32LB520-3-5-7-HW-Application.html
https://wiki.sifli.com/en/hardware/SF32LB52B-E-G-J-HW-Application.html

0-3-5-7 are powered using a Lithium battery and support USB charging.
B-E-G-J are powered at 3.3V and do not support charging.

Other termination codes indicate what type of memory and size is
embedded in the package (QSPI NOR or PSRAM).

Other families exist within the SF32LB family, like SF32LB56x,
SF32LB58x, etc.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: add skeleton for sf32lb52x

Just define basic ARM hardware: CPU, MPU and NVIC settings. SF32LB SoCs
are technically Arm-Star MC1 based, an Arm Cortex-M33 compatible CPU
developed by Arm China.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: add sram layout default schemes

SRAM is segmented as:

- RAM0: 128K (DTCM)
- RAM1: 128K
- RAM2: 256K

where:

- RAM0 can be accessed by all peripherals except PTC1/2.
- RAM1/RAM2 have its own port, so e.g. master 1 can access RAM1 while
  master 2 can access RAM2.

Depending on the application, one may decide to just use the whole RAM
as a single block, split RAM0 (DTCM) and RAM1/2, etc. For now, provide 2
schemes:

- Use all SRAM as a single block
- Use RAM0 (DTCM) + RAM1/2

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: sifli: add sf32lb52_devkit_lcd

Ref. https://wiki.sifli.com/en/board/sf32lb52x/SF32LB52-DevKit-LCD.html

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
scripts: west_commands: runners: add sftool runner

So we can flash using sftool.

Ref. https://github.com/OpenSiFli/sftool.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: common: add sftool runner support

For the recently added sftool west runner.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: add puya vendor prefix

To be used for some NOR flash devices.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: memory-controllers: add sifli,sf32lb-mpi

Binding for SiFli SF32LB Memory Peripheral Interface (MPI).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define mpi1/2

MPI1/2 are memory controllers (PSRAM, NOR, NAND, etc.).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
manifest: add hal_sifli

Add SiFli HAL module. This module contains register-level description
for SF32LB SoCs IP, and a high-level HAL.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: hal_sifli: add initial glue code

Add glue code to build SiFli HAL. The bare minimum is included, some
more things may be required as we fully test the integration in the
future.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: sifli: sf32: signal HAL selection and use CMSIS HAL glue code

This brings CMSIS HAL glue code, which is required if we want to enable
any HAL module without everything blowing up. cmsis_core_m_defaults.h
cannot be used once we enable HAL. CMSIS is that nice.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: serial: add sifli,sf32lb-usart

For the UART IP in SF32LB SoCs.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define usart1

Define node for USART1. Other available USART instances will be added in
the future.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: serial: sf32lb: add basic driver

Just basic driver with poll-in/out capability.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: pinctrl: add sifli,sf32lb52x-pinmux

Add bindings for the SF32LB52X SoCs PINMUX peripheral. Note that
SF32LB56X SoCs contain a compatible IP (not others, where HPSYS_CFG is
not required), so if SF32LB56X support is added, binding could be
adjusted to reflect 52x/56x support.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define pinctrl

Add a node for pinctrl (PINMUX).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dt-bindings: pinctrl: add pin definitions boilerplate for SF32LB52x

Add valid pin configurations boilerplate for SF32LB52x. All valid
configurations will be added in a follow-up commit.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: pinctrl: sf32lb52x: initial driver

Initial driver for SF32LB52X SoCs.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: syscon: add sifli,sf32lb-cfg

SiFli SoCs have a multi-purpose register set named HPSYS_CFG, which we
can treat as a "syscon".

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define cfg

Define node for SF32LB HPSYS_CFG.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define all fixed clocks

Refer to Ch 2 "Clock and Reset" from reference manual. Some may require
specific bindings in the future if they can be configured (e.g. RC48).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: mfd: add sifli,sf32lb-rcc

SiFli SF32LB RCC (Reset and Clock Control).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define RCC

Add node for RCC.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: clock: add sifli,sf32lb-rcc-clk

Add binding for SF32LB RCC (clock part).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define RCC clock controller

This is defined as a subnode because RCC is a MFD device (clock and
reset controller), however, Zephyr does not allow >1 device per DT node!

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dt-bindings: clock: add definitions for SF32LB52X clocks

So they can easily be referenced in devicetree.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: clock_control: sf32lb: add initial driver

Initial driver for SF32LB SoCs. This driver is incomplete, but allows
to configure the system for a minimal boot.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: clock: add sifli,sf32lb-dll

Add binding for SF32LB DLL clock.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
MAINTAINERS: add SiFli SF32LB entry

Add myself as a maintainer for now, and @cameled as a collaborator.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
MAINTAINERS: add hal_sifli entry

Add myself as a maintainer for now, and @cameled as a collaborator.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

nordicjm commented on boards/sifli/sf32lb52_devkit_lcd/sf32lb52_devkit_lcd-pinctrl.dtsi:
```diff
@@ -0,0 +1,19 @@
+		group1 {
+			pinmux = <PA19_USART1_TXD>;
+		};
+		group2 {
```
newline line 13

nordicjm replied:
as per other commit

nordicjm commented on boards/sifli/sf32lb52_devkit_lcd/doc/img/sf32lb52-devkit-lcd.webp:
(optional) board resolution is large, could shrink to reduce file size

nordicjm reviewed (changes requested):
some minor nit. board/soc files OK other than listed issues

gmarull committed:
dts: bindings: power: add sifli,sf32lb-aon

SiFli SF32LB AON module controls certain low-power or clocks, e.g.
enablement of HXT48 on different LP modes, etc.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define AON entry

Add entry for SF32LB AON.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: clock: add sifli,sf32lb-hxt48

Add binding for SF32LB HXT48 clock.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define HXT48 clock

Add node for the HXT48 clock.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: clock_control: sf32lb_hxt48: add initial driver

Add driver for HXT48.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: power: add sifli,sf32lb-pmuc

Add bindings for the SiFli SF32LB PMUC. PMUC is a sort of "syscon" type
register block used to control multiple power-management related stuff.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: sifli: sf32lb52x: define pmuc

Add a devicetree node for PMUC.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

fabiobaltieri commented on dts/bindings/power/sifli-sf32lb-pmuc.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2025 Core Devices LLC
+# SPDX-License-Identifier: Apache-2.0
+
+
```
same

fabiobaltieri commented on include/zephyr/drivers/clock_control/sf32lb.h:
```diff
@@ -0,0 +1,104 @@
+ * @param node_id DT node identifier
+ * @param default_spec Default value to be used if the node has no clocks property.
+ */
+#define SF32LB_CLOCK_DT_SPEC_GET_OR(node_id, default_spec)                                         \
+	COND_CODE_1(DT_CLOCKS_HAS_IDX(node_id, 0),                                        \
```
nit, `\` are off

fabiobaltieri reviewed (commented):
Good stuff, just a couple nitpicks really

gmarull commented on soc/sifli/sf32/sf32lb52x/soc.h:
```diff
@@ -6,6 +6,6 @@
 #define _SOC_SIFLI_SF32_SF32LB52X_SOC_H_
 
-#include <cmsis_core_m_defaults.h>
+#include <register.h>
```
nope, it's intended to make things bisectable. register.h is brought by HAL, and contains the CMSIS IRQn typedeffery which without HAL is in cmsis_core_m_defaults.h.

gmarull replied:
If you want to have fun: https://github.com/zephyrproject-rtos/zephyr/issues/57246

cameled reviewed (commented):
Test on `SF32LB52-DevKit-LCD` board.
```
SFBL
    *** Booting Zephyr OS build v4.2.0-2975-g142fdd96eb30 ***
Hello World! sf32lb52_devkit_lcd/sf32lb525uc6
```

gmarull commented on scripts/west_commands/runners/sftool.py:
```diff
@@ -0,0 +1,86 @@
+            "--port",
+            type=str,
+            required=True,
+            help="Serial port device, e.g. /dev/ttyUSB0",
```
that's expected, you need to run `west flash --context` to print these

gmarull replied:
with `west flash` you'd get:

```
(.venv) ➜  zephyr git:(sf32lb52x) west flash          
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner sftool
usage: sftool [-h] [-d DIR] [-r RUNNER] [--skip-rebuild] [--domain DOMAIN] [-H] [--board-dir DIR] [--gdb GDB] [--openocd OPENOCD] [--openocd-search DIR]
              [--dt-flash {Y,y,N,n,yes,no,YES,NO}] [-f FILE] [-t FILE_TYPE] [--elf-file FILE] [--hex-file FILE] [--bin-file FILE] [--mot-file FILE] [--erase]
              [-O TOOL_OPT] --chip CHIP --port PORT
sftool: error: the following arguments are required: --port
```

fabiobaltieri reviewed (dismissed):
Good stuff, guess this needs https://github.com/zephyrproject-rtos/zephyr/issues/95478 to be sorted out first but looks good otherwise.

nordicjm commented on modules/hal_sifli/CMakeLists.txt:
```diff
@@ -0,0 +1,160 @@
+      CONFIG_USE_SIFLI_HAL_WDT ${ZEPHYR_HAL_SIFLI_MODULE_DIR}/hal/src/bf0_hal_wdt.c
+    )
+
+    if (CONFIG_USE_SIFLI_HAL_AES OR CONFIG_USE_SIFLI_HAL_HASH)
```
no space between if and bracket in cmake

nordicjm commented on soc/sifli/sf32/Kconfig.soc:
```diff
@@ -0,0 +1,11 @@
+
+config SOC_FAMILY_SF32
+	bool
+	select HAS_SIFLI_HAL
```
```suggestion
```

nordicjm commented on soc/sifli/sf32/Kconfig:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_FAMILY_SF32
+
```
```suggestion

config SOC_FAMILY_SF32
	select HAS_SIFLI_HAL

```
